libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt

```diff
@@ -411,4 +423,20 @@ internal object ManeuverProcessor {
         }
         return componentList
     }
+
+    private fun removeDuplicateManeuvers(
+        originalList: List<Maneuver>
+    ): List<Maneuver> {
+        return originalList.fold(mutableListOf()) { currentList, currentItem ->
+            if (currentList.isEmpty()) {
+                mutableListOf(currentItem)
+            } else {
+                if (currentItem.equalsIgnoreId(currentList.last())) {
+                    currentList
+                } else {
+                    currentList.apply { add(currentItem) }
+                }
```

Comment:
Do these banners have different distance remaining values? If yes, doesn't that break this logic?

Reply:
Are you saying the logic will break because the distance remaining values will be different?

Reply:
Yes, that's my question - if we have 2 banner instructions for a single step, will they have different distances? Doesn't that break the comparison?

Reply:
Yes that's exactly what is happening
#### Before
<img width="250" src="https://user-images.githubusercontent.com/9770186/124891215-65ee5880-dff6-11eb-8fd1-e0f32062db7f.png">

The banner instruction with the text `Stevenson Boulevard` are 2 instructions in a single step. If you see the upcoming list, all we care about only is the instruction since the current banner shows distance remaining and upcoming banner shows total distance. I have added `Maneuver.equalsIgnoreId()` that compares only `PrimaryManeuver` and `SecondaryManeuver` without the `id` and `stepDistance` because both id and stepDistance is going to be different.

#### After
<img width="250" src="https://user-images.githubusercontent.com/9770186/124891372-8cac8f00-dff6-11eb-9822-abc61c6244cf.png">

Reply:
Now that I think about it, why are we even using a comparison here?

We do have the list of banners for each step in the route, why don't we just always take one banner from that list instead of taking all of them and then filtering out? The banners in a list will always point to one and the same maneuver anyway, so it doesn't matter which one we pick, does it?

Reply:
> We do have the list of banners for each step in the route

That's not entirely true. Every single step will not have more than one banner instruction.

> why don't we just always take one banner from that list instead of taking all of them and then filtering out?

Lets say you have the following list of instruction in a particular step
1: Maneuver(primaryManeuver = "Stevenson Boulevard", secondaryManeuver = null, subManeuver = "StevensonSt")
2: Maneuver(primaryManeuver = "Stevenson St", secondaryManeuver = I-880 North, subManeuver = "Laurel Place") 

How will you decide which maneuver to take and which one to leave? If we leave one of them, when that instruction becomes current instruction, we will miss out the sub banner and secondary banner. We could have done what you are suggesting if we only have the primary maneuver.

cc @LukasPaczos

Reply:
Your scenario assumes that 2 instructions in a step are different, is that actually ever the case? Aren't instructions in a single step always describing the same maneuver?

Do you know how does this logic work on iOS? I feel like we're trying to reinvent the wheel.

Reply:
> Your scenario assumes that 2 instructions in a step are different, is that actually ever the case

That is the case most of the time. If there are 2 instructions in a step, the last instruction would most of the time have a `sub` attached to it, because that is going to be the primary banner in the next step. Look at the same route I have used in the screenshot

```
          "steps": [
            {
              "bannerInstructions": [
                {
                  "primary": {
                    "components": [
                      {
                        "type": "text",
                        "text": "Stevenson Boulevard"
                      }
                    ],
                    "type": "turn",
                    "modifier": "uturn",
                    "text": "Stevenson Boulevard"
                  },
                  "distanceAlongGeometry": 1095.189
                },
                {
                  "sub": {
                    "components": [
                      {
                        "imageBaseURL": "https://mapbox-navigation-shields.s3.amazonaws.com/public/shields/v4/US/i-880",
                        "type": "icon",
                        "text": "I 880"
                      },
                      {
                        "type": "text",
                        "text": "North"
                      }
                    ],
                    "type": "off ramp",
                    "modifier": "right",
                    "text": "I 880 North"
                  },
                  "primary": {
                    "components": [
                      {
                        "type": "text",
                        "text": "Stevenson Boulevard"
                      }
                    ],
                    "type": "turn",
                    "modifier": "uturn",
                    "text": "Stevenson Boulevard"
                  },
                  "distanceAlongGeometry": 402.336
                }
              ]
```

https://api.mapbox.com/directions/v5/mapbox/driving-traffic/-121.981985,37.529766;-121.98651,37.522633?access_token=<your_access_token>&geometries=polyline6&overview=full&steps=true&continue_straight=true&annotations=congestion%2Cmaxspeed%2Cspeed%2Cduration%2Cdistance%2Cclosure&language=en&roundabout_exits=true&voice_instructions=true&banner_instructions=true&voice_units=imperial&enable_refresh=true

Reply:
> Do you know how does this logic work on iOS? I feel like we're trying to reinvent the wheel.

I have no idea how iOS is doing it.

Reply:
There's another element to this puzzle that I think we're currently missing, which is timing - if only the second instruction has sub banner, it intentionally delays when the sub is shown. If we filtered one of the instructions out in the data (or merged them together, or only used the last element), we would lose that timing information.

I think the _API_ (that prepares the data) cannot be responsible for filtering here. The filtering should happen in the presentation layer.

And if we did move the filtering to the presentation layer, I see the logic working like this:
1. View receives the data
2. Step has multiple banners
  1. If it's the current step, use the first banners in the list (API already prepared this) and filter out all thy other banners
  2. If it's not the current step, filter out all banners besides last for the given step

The only piece of information that is probably needed to pull this off is adding `stepIndex` information to each `Maneuver` object.

Reply:
Discussed internally that timing is not an issue here and it is preserved with how the list of maneuvers is transformed on each route progress update :+1: 

We'll also need to do a similar transformation when route progress is not available and just take the first instruction for each step @abhishek1508.